notify/jira/jira.go

```diff
@@ -0,0 +1,319 @@
+// Copyright 2023 Prometheus Team
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package jira
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	commoncfg "github.com/prometheus/common/config"
+	"github.com/prometheus/common/model"
+	"github.com/trivago/tgo/tcontainer"
+
+	"github.com/prometheus/alertmanager/config"
+	"github.com/prometheus/alertmanager/notify"
+	"github.com/prometheus/alertmanager/template"
+	"github.com/prometheus/alertmanager/types"
+)
+
+// Notifier implements a Notifier for JIRA notifications.
+type Notifier struct {
+	conf    *config.JiraConfig
+	tmpl    *template.Template
+	logger  log.Logger
+	client  *http.Client
+	retrier *notify.Retrier
+}
+
+func New(c *config.JiraConfig, t *template.Template, l log.Logger, httpOpts ...commoncfg.HTTPClientOption) (*Notifier, error) {
+	client, err := commoncfg.NewClientFromConfig(*c.HTTPConfig, "jira", httpOpts...)
+	if err != nil {
+		return nil, err
+	}
+	return &Notifier{
+		conf:    c,
+		tmpl:    t,
+		logger:  l,
+		client:  client,
+		retrier: &notify.Retrier{RetryCodes: []int{http.StatusTooManyRequests}},
+	}, nil
+}
+
+// Notify implements the Notifier interface.
+func (n *Notifier) Notify(ctx context.Context, as ...*types.Alert) (bool, error) {
+	key, err := notify.ExtractGroupKey(ctx)
+	if err != nil {
+		return false, err
+	}
+
+	level.Debug(n.logger).Log("alert", key)
+
+	var (
+		tmplTextErr error
+
+		alerts       = types.Alerts(as...)
+		data         = notify.GetTemplateData(ctx, n.tmpl, as, n.logger)
+		tmplText     = notify.TmplText(n.tmpl, data, &tmplTextErr)
+		tmplTextFunc = func(tmpl string) (string, error) {
+			result := tmplText(tmpl)
+			return result, tmplTextErr
+		}
+	)
+
+	existingIssue, shouldRetry, err := n.searchExistingIssue(key, alerts.Status())
+	if err != nil {
+		return shouldRetry, fmt.Errorf("error searching existing issues: %w", err)
+	}
+
+	// Do not create new issues for resolved alerts
+	if existingIssue == nil && alerts.Status() == model.AlertResolved {
+		return false, nil
+	}
+
+	path := "issue"
+	method := http.MethodPost
+
```

Comment:
To make it clear to read, why do not move 
```
	path := "issue"
	method := http.MethodPost
```
into `if existingIssue == nil {`
you are redefining in `else` condition